exporters/metric/cortex/auth_test.go

```diff
@@ -153,3 +164,353 @@ func createFile(bytes []byte, filepath string) error {
+	require.Nil(t, err)
+	defer os.Remove("./ca_cert.pem")
+	defer os.Remove("./ca_key.pem")
+
+	// Generate certificate for the server. The client will check this certificate against
+	// its certificate authority to verify the server.
+	_, _, err = generateServingCertFiles(
+		caCert,
+		caPrivateKey,
+		"./serving_cert.pem",
+		"./serving_key.pem",
+	)
+	require.Nil(t, err)
+	defer os.Remove("./serving_cert.pem")
+	defer os.Remove("./serving_key.pem")
+
+	// Generate certificate for the client. The server will check this certificate against
+	// its certificate authority to verify the client.
+	_, _, err = generateClientCertFiles(
+		caCert,
+		caPrivateKey,
+		"./client_cert.pem",
+		"./client_key.pem",
+	)
+	require.Nil(t, err)
+	defer os.Remove("./client_cert.pem")
+	defer os.Remove("./client_key.pem")
+
+	// Generate the tls Config to set up mutual TLS on the server.
+	serverTLSConfig, err := generateServerTLSConfig(
+		"ca_cert.pem",
+		"serving_cert.pem",
+		"serving_key.pem",
+	)
+	require.Nil(t, err)
+
+	// Create and start the TLS server.
+	handler := func(rw http.ResponseWriter, req *http.Request) {
+		fmt.Fprint(rw, "Successfully verified client and received request!")
+	}
+	server := httptest.NewUnstartedServer(http.HandlerFunc(handler))
+	server.TLS = serverTLSConfig
+	server.StartTLS()
+	defer server.Close()
+
+	// Create an Exporter client with the client and CA certificate files.
+	exporter := Exporter{
+		Config{
+			TLSConfig: map[string]string{
+				"ca_file":              "./ca_cert.pem",
+				"cert_file":            "./client_cert.pem",
+				"key_file":             "./client_key.pem",
+				"insecure_skip_verify": "0",
+			},
+		},
+	}
+	client, err := exporter.buildClient()
+	require.Nil(t, err)
+
+	// Send the request and verify that the request was successfully received.
+	res, err := client.Get(server.URL)
+	require.Nil(t, err)
+	defer res.Body.Close()
+}
+
+// generateCertFiles generates new certificate files from a template that is signed with
+// the provided signer certificate and key.
+func generateCertFiles(
+	template *x509.Certificate,
+	signer *x509.Certificate,
+	signerKey *rsa.PrivateKey,
+	certFilepath string,
+	keyFilepath string,
+) (*x509.Certificate, *rsa.PrivateKey, error) {
+	// Generate a private key for the new certificate. This does not have to be rsa 4096.
+	privateKey, err := rsa.GenerateKey(rand.Reader, 4096)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Check if a signer key was provided. If not, then use the newly created key.
+	if signerKey == nil {
+		signerKey = privateKey
+	}
+
+	// Create a new certificate in DER encoding.
+	encodedCert, err := x509.CreateCertificate(
+		rand.Reader, template, signer, privateKey.Public(), signerKey,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Write the certificate to the local directory.
+	certPEM := pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: encodedCert,
+	})
+	err = createFile(certPEM, certFilepath)
```

Comment:
definitely a nit, but having this seems pretty circuitous given the [`Encode`](https://golang.org/pkg/encoding/pem/#Encode) function from the `pem` module accepts an `io.Writer`. Instead of writing everything to memory it could be directly written to a file handle if this was restructured.

Pretty sure it's not worth spending time changing given this is for a testing file, but just wanted to encourage the change if you were already on the fence. :wink:

Reply:
Did not notice that about the `Encode` function! I adjusted it in the recent commits.